PhysicsTools/PatAlgos/plugins/GenJetFlavourInfoPreserver.cc

```diff
@@ -66,26 +64,19 @@ pat::GenJetFlavourInfoPreserver::produce(edm::Event & iEvent, const edm::EventSe
     Handle<reco::JetFlavourInfoMatchingCollection> genJetFlavourInfos;
     iEvent.getByToken(genJetFlavourInfosToken_, genJetFlavourInfos);
 
-    Handle<std::vector < std::pair<  int, int > > > slimmedGenJetAssociation;
+    Handle<edm::Association<std::vector<reco::GenJet> > > slimmedGenJetAssociation;
     iEvent.getByToken(slimmedGenJetAssociationToken_, slimmedGenJetAssociation);
 
     auto jetFlavourInfos = std::make_unique<reco::JetFlavourInfoMatchingCollection>(reco::JetRefBaseProd(slimmedGenJets));
-    auto jetInfoIt = genJetFlavourInfos->begin(), jetInfoEnd = genJetFlavourInfos->end();
     assert(genJets->size() == genJetFlavourInfos->size());
 
-    auto assoIt = slimmedGenJetAssociation->begin(), assoEnd = slimmedGenJetAssociation->end();
-
-    for(; assoIt != assoEnd; ++assoIt){
-        for (; jetInfoIt != jetInfoEnd; ++jetInfoIt ){
-            if ((jetInfoIt-genJetFlavourInfos->begin()) == assoIt->first){
-                (*jetFlavourInfos)[slimmedGenJets->refAt(assoIt->second)] = reco::JetFlavourInfo(jetInfoIt->second.getbHadrons(),
-                                                                                    jetInfoIt->second.getcHadrons(),
-                                                                                    jetInfoIt->second.getPartons(), 
-                                                                                    jetInfoIt->second.getLeptons(), 
-                                                                                    jetInfoIt->second.getHadronFlavour(), 
-                                                                                    jetInfoIt->second.getPartonFlavour());
-            }
-         }
+    edm::Ref<std::vector<reco::GenJet> > slimmedGenJetRef;
+
+    for( edm::View<reco::GenJet>::const_iterator genJetsIt = genJets->begin(); genJetsIt != genJets->end(); ++genJetsIt){
+        slimmedGenJetRef = (*slimmedGenJetAssociation)[genJets->refAt(genJetsIt-(genJets->begin()))];
```

Comment:
we are getting close. The following is going to make the loop implementation less verbose
```
int i = -1;
for (auto const& genJet : *genJets){
         i++;
        slimmedGenJetRef = (*slimmedGenJetAssociation)[genJets->refAt(i)]; 
        if(!slimmedGenJetRef) continue;
        (*jetFlavourInfos)[reco::JetBaseRef(slimmedGenJetRef)] = (*genJetFlavourInfos)[i].second;
}
```